xrdp/xrdp_painter.c

```diff
@@ -94,47 +99,145 @@ xrdp_painter_send_dirty(struct xrdp_painter *self)
 
     LLOGLN(10, ("xrdp_painter_send_dirty:"));
 
+    int width = self->wm->screen->width;
+    int height = self->wm->screen->height;
+
     bpp = self->wm->screen->bpp;
     Bpp = (bpp + 7) / 8;
     if (Bpp == 3)
     {
         Bpp = 4;
     }
 
-    jndex = 0;
-    error = xrdp_region_get_rect(self->dirty_region, jndex, &rect);
-    while (error == 0)
+    if (self->wm->codec_handle)
     {
-        cx = rect.right - rect.left;
-        cy = rect.bottom - rect.top;
-        ldata = (char *)g_malloc(cx * cy * Bpp, 0);
-        if (ldata == 0)
+        int num_dirty = xrdp_region_get_numrects(self->dirty_region);
+        if (num_dirty == 0)
         {
-            return 1;
+            return 0;
         }
-        src = self->wm->screen->data;
-        src += self->wm->screen->line_size * rect.top;
-        src += rect.left * Bpp;
-        dst = ldata;
-        for (index = 0; index < cy; index++)
+
+        int tiles_wide = (width + RFX_TILE_SQUARE - 1) / RFX_TILE_SQUARE;
+        int tiles_high = (height + RFX_TILE_SQUARE - 1) / RFX_TILE_SQUARE;
+        char tile_map[tiles_wide][tiles_high];
+        g_memset(tile_map, 0, sizeof(tile_map));
+
+        struct rfx_rect rfxrects[num_dirty];
+        struct rfx_tile rfxtiles[tiles_wide * tiles_high];
+        int num_rfxrects = 0;
+        int num_tiles = 0;
+
+        jndex = 0;
+        error = xrdp_region_get_rect(self->dirty_region, jndex, &rect);
+        while (error == 0)
         {
-            g_memcpy(dst, src, cx * Bpp);
-            src += self->wm->screen->line_size;
-            dst += cx * Bpp;
+            rfxrects[num_rfxrects].x = rect.left;
+            rfxrects[num_rfxrects].y = rect.top;;
+            rfxrects[num_rfxrects].cx = rect.right - rect.left;
+            rfxrects[num_rfxrects].cy = rect.bottom - rect.top;
+            num_rfxrects++;
+
+            int top_ty = rect.top / RFX_TILE_SQUARE;
+            int bot_ty = (rect.bottom - 1) / RFX_TILE_SQUARE;
+            for (int ty = top_ty; ty <= bot_ty; ty++)
+            {
+                int left_tx = rect.left / RFX_TILE_SQUARE;
+                int right_tx = (rect.right - 1) / RFX_TILE_SQUARE;
+                for (int tx = left_tx; tx <= right_tx; tx++)
+                {
```

Comment:
For loop initial declarations are only allowed in C99 -- not otherwise required by xrdp.